processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingOptions.java

```diff
@@ -54,13 +58,171 @@ public MappingOptions(Map<String, List<Mapping>> mappings, IterableMapping itera
      */
     public Map<String, List<Mapping>> getMappings() {
         return mappings;
     }
 
+    /**
+     * The target references are popped. The MappingOptions are keyed on the unique first entries of the
+     * target references.
+     *
+     * So, take
+     *
+     * targetReference 1: propertyEntryX.propertyEntryX1.propertyEntryX1a
+     * targetReference 2: propertyEntryX.propertyEntryX2
+     * targetReference 3: propertyEntryY.propertyY1
+     * targetReference 4: propertyEntryZ
+     *
+     * will be popped into:
+     *
+     * propertyEntryX - MappingOptions ( targetReference1: propertyEntryX1.propertyEntryX1a,
+     *                                   targetReference2: propertyEntryX2 )
+     * propertyEntryY - MappingOptions ( targetReference1: propertyEntryY1 )
+     *
+     * So, 2 cloned new MappingOptions with popped targetReferences. Also Note that the not nested targetReference4
+     * disappeared.
+     *
+     * @return See above
+     */
+    public Map<PropertyEntry, MappingOptions> popTargetReferences() {
+
+        // sort all mappings based on the top level name before popping
+        Map<PropertyEntry, List<Mapping>> mappingsKeyedByProperty = new HashMap<PropertyEntry, List<Mapping>>();
+        for ( List<Mapping> mapping : mappings.values() ) {
+            Mapping newMapping = first( mapping ).popTargetReference();
+            if ( newMapping != null ) {
+                // sort properties on current name.
+                PropertyEntry property = first( first( mapping ).getTargetReference().getPropertyEntries() );
+                if ( !mappingsKeyedByProperty.containsKey( property ) ) {
+                    mappingsKeyedByProperty.put( property, new ArrayList<Mapping>() );
+                }
+                mappingsKeyedByProperty.get( property ).add( newMapping );
+            }
+        }
+
+        // now sort them into mapping options
+        Map<PropertyEntry, MappingOptions> result = new HashMap<PropertyEntry, MappingOptions>();
+        for (  Map.Entry<PropertyEntry, List<Mapping>> mappingKeyedByProperty : mappingsKeyedByProperty.entrySet() ) {
+            Map<String, List<Mapping>> newEntries = new HashMap<String, List<Mapping>>();
+            for ( Mapping newEntry : mappingKeyedByProperty.getValue() ) {
+                newEntries.put( newEntry.getTargetName(), Arrays.asList( newEntry ) );
+            }
+            result.put( mappingKeyedByProperty.getKey(), forMappingsOnly( newEntries ) );
+        }
+        return result;
+    }
+
+    /**
+     * Check there are nested target references for this mapping options.
+     *
+     * @return boolean, true if there are nested target references
+     */
+    public boolean hasNestedTargetReferences() {
+        boolean found = false;
+        Iterator<List<Mapping>> mapIterator = mappings.values().iterator();
+        while ( mapIterator.hasNext() && !found ) {
```

Comment:
Fine by me. I just needed some more time to understand the loops, nothing else 😄